### ISSUE TITLE
`id` as default locator

### ISSUE THREAD

**Issue by pekkaklarck** [MarketSquare/robotframework-browser#1143]
Browser apparently uses CSS selectors as locators by default. That's a good idea, but it would be convenient if element ids could be used directly as well. There are many different reasons why I think it would make sense:

- `#` is a comment character in RF data so CSS selector like `#example` requires escaping like `\#example`. This is a bit ugly and the syntax may not be familiar for everyone.
- Many (most?) elements to interact with (input fields, buttons, ...) have ids so direct support would be convenient. 
- SeleniumLibrary supports ids by default. Browser supporting them would ease transition. SeleniumLibrary also supports name attribute by default but that's not so relevant nowadays.

I would assume the implementation could be something like this:
1. Check does the locator have explicit prefix specifying the type. If yes, just use whatever locator specified.
2. Check does the locator start with `//`. If yes, it's an xpath expression.
3. Check can an element be found by using the locator as an id. If yes, go with it.
4. Use the locator as a CSS selector.

I believe the current implementation is like above but without step 3. Checking is there an element by an id ought to be so fast that this extra step shouldn't matter. Explaining this special behavior shouldn't be too complicated either. I can write documentation, with examples, if needed. :smiley:

**Comment by mkorpela** [comment]
If I understand this correctly the problem is that # is a special comment char in RF.
Problem situation occurs when a user tries to directly use that css id selector without escaping #

**Comment by pekkaklarck** [comment]
Yes. I'd say SeleniumLibrary compatibility would be equally important. I know Browser doesn't try to be drop-in replacement, but compatibility where possible would nevertheless ease transition. Ids just work with SL, I was personally surprised that wasn't the case with B.

**Comment by mkorpela** [comment]
Basically 2 problems:
1. `#` being a special char
2. SeleniumLibrary having differing default selector strategy ( [id and name + case-by-case something else](https://robotframework.org/SeleniumLibrary/SeleniumLibrary.html#Default%20locator%20strategy) )

**Comment by pekkaklarck** [comment]
Yeah. I don't think there's a need to emulate SLs default locator strategy fully. I doubt `name` is anymore that widely used and searching by `name` and `id` is likely more complicated and slower than searching by `id` only.

**Comment by mkorpela** [comment]
We are bound pretty much by Playwright choices on selector strategies. Basically only thing we have done is automatic frame piercing because PW did not want to implement this for the asynchronousity reasons.. and in JS land that is in my opinion the sensible default BUT in synchronous RF land frame piercing is fine.
PW selector strategies have evolved during last year and they might do so also in the future.
Also a selector strategy fails only after timeout, this would be a big change to support multiple defaults or extremely time consuming during test execution.
For these reasons, I feel that Id default strategy is a risk and might be maintenance problem in the future.
I think it might be better to check existence of a matching ID if failing in selecting and reporting the matching selector + how to correctly write it in the failure message.

**Comment by mkorpela** [comment]
comment vs `#` is currently failing also already in RF side, so library can not notify user about this before RF already states that not enough params given for the keyword

**Comment by mkorpela** [comment]
Could RF actually give a suggestion to the user to escape # in case Keyword has not enough arguments and has a comment? @pekkaklarck

**Comment by aaltat** [comment]
I agree with Mikko about the locator strategy. Way Playwright works is very different than way Selenium works. In Selenium it is easier to build custom functionality, because Selenium is pretty flexible how locating elements works. Selenium API is pretty flexible on this matter and offers many ways. Therefore it is easy to build functionality in the SL level and make it work one wants in the SL  level. But in PW side, things are more hard coded and we took the way that we do not mangle the locator in any way. We just pass it in as is and let the Playwright to do handling. That way we are not tied tightly to PW version when locator strategy evolves.

**Comment by pekkaklarck** [comment]
I thought PW would have a method like `getElementById` that could be used for a *fast* check does an element exist or not. Apparently it doesn't and that makes implementing this considerably harder. It would be possible to just add `id=` prefix to the given locator and pass it to PW, but that has its problems as well. There would be two options how to handle this:

- First try with the `id=` prefix and then without it. If only the latter works, the time spent for the first attempt is all waste. Due to timeouts there can be considerable amount of time wasted so this is probably not a good idea.
- Only try with the `id=` prefix unless the user indicates that some other locator strategy should be used. This is close to how SL handles this (also looking for `name` is pretty irrelevant) and I personally like that approach. Changing this would have serious problems, though:
  - Backwards incompatible. Tests using PW's default locator strategy would fail.
  - Same examples as with plain PW wouldn't work anymore.
  - CSS as the default strategy has plenty of benefits itself.

Due to both of the above options having severe problems, changing this is most likely not a good idea. What could be done instead:

- Mention differences to SL locator strategy in Browser keyword docs. No need to go through all the details, but explaining main differences (especially that the default strategy is different) could help people migrating from SL.
- Support `:` as the locator strategy separator. This would allow usages like `id:example` that would work the same way with SL and B. Probably more importantly, `name:value` syntax doesn't clash with Robot's named argument syntax `name=value`. This was the reason SL got the support for `:` few years ago (robotframework/SeleniumLibrary#908).

**Comment by pekkaklarck** [comment]
RF being able to warn about incorrect `#` usage would be rather hard to implement. At the moment the execution side model doesn't know anything about comments and adding this info would be very big task. I doubt it's worth the effort, especially when editors already now show comments using different color.

My main problem with `\#example` is that it's a bit ugly. Alternative of using `id=example` is problematic due to looking like the named argument syntax. As I commented above, I'd like `id:example` more (and can submit an issue about that if you think that's fine and feasible to implement). Anyway, locators shouldn't be used on the test level so somewhat ugly usages like `\#example` aren't a big problem.

**Comment by adrian-evo** [comment]
In my tests I am using `id=example` but also `text=example` and I agree it looks like named argument. So I would vote for `id:example`, and `text:example`.

**Comment by mkorpela** [comment]
I'm sorry but that choice goes deep into Playwright selector / locator parsing. I'm afraid that diverting from the basic Playwright selector parsing will be hard especially now that there is some mass of tests already done with Browser library.
We had issues with `text` getting mixed with argument name and I agree with the concerns related.

**Comment by aaltat** [comment]
In Selenium one first finds the element and something with the element. Something like:
```
element  = driver.find_element_by_name("q")
element.click()
```
Almost all locator strategies in SeleniumLibrary side are different API calls in Selenium side. Now everything, but many. To simplify with users, SL maps many things as locator strategies.

But in [PW side](https://playwright.dev/docs/selectors), things works differently, on page do something with the element: 
```
await page.click('id=q');
``` 
PW has a locator strategy, which already uses `=` and `:` characters for different kind of meanings. Looking at the current PW docs, it might be possible to support `:` as locator strategy separator, at least with quick look I do not see anything that would be impossible to solve. But it would require lot more testing for us in the library side, so that we keep the library transformation compatible with PW side. For now library just passes locator string as in to PW side and lets PW to handle the locator resolving. 

Yes it is an easy choice for us, but also simplifies code and specially documentation in our side. It would difficult to explain to people that in some case, library will do some conversion and in some case it is not converted and passed as in to PW side. It would be hard to see, when something is done in library side and when something is done in PW side. I agree that `text=something` is not pretty in RF side, but think that it is better than trying to explain users bunch of exceptions to PW docs.

**Comment by mkorpela** [comment]
We have discussed about this with Robot Framework Browser team and decision was that we will:
- improve documentation about id selectors
- improve documentation about differences with SL and Browser

`id` as a default locator was decided not to be done.
Use of id:s has greatly decreased in web development during the years.
Most of stuff are done with css classes. React (or alternatives) kind of handle the part where you would have need id before.
Basically cases where a value in a field needs to be updated and that field needs to be located.
This means that newer web applications do not so much relay on use of `id`s.
CSS classes on the other hand are used quite a lot as targeting in styling of components. They also now are in my opinion the base of the most stable selector strategies.

For these reasons and discussions before, I'm now closing this issue.
